In [126]:
from difflib import SequenceMatcher

def similar(a, b):
    return SequenceMatcher(None, a, b).ratio()

In [127]:
def comparar2(cadena1, cadena2):
    
    cadena1 = cadena1.replace(".", ". ").replace("  ", " ").strip().replace(",", "")
    cadena2 = cadena2.replace(".", ". ").replace("  ", " ").strip().replace(",", "")

    l1, l2 = cadena1.split(" "), cadena2.split(" ")
    if len(cadena1)<len(cadena2):
        fd = l1
        l1 = l2
        l2 = fd
    
    if len(l1)==len(l2) and len(l1)>1:
        verdad = []
        unos = []
        for i in range(len(l1)):
            lett1 = l1[i][0]
            lett2 = l2[i][0]
            uno = similar(l1[i], l2[i])
            if lett1==lett2:
                verdad.append(True)
            if uno==1.0:
                unos.append(uno)
                
        if sum(verdad)==len(l1):
            if len(l1)==2 and sum(unos)>=1.8:
                return True
            elif len(l1)>2 and sum(unos)>=len(l1)-1:
                return True
            else:
                return False
        else:
            return False
            
            
    elif len(l1)==3 and len(l2)==2 or len(l1)==2 and len(l2)==3:
        lett1_1 = l1[0][0]
        lett2_1 = l2[0][0]

        lett1_2 = l1[-1][0]
        lett2_2 = l2[-1][0]
        
        verdad = [0, 0]
        
        indicador = 0
        for i in l1:
            for j in l2:
                nn = similar(i, j)
                if nn>=0.8:
                    indicador+=nn
        if lett1_1==lett2_1 and lett1_2==lett2_2 and indicador >= len(l1)-1:
            return True
        else:
            return False
        
        
    elif len(l1)>3 and len(l2)>=3 or len(l1)>=3 and len(l2)>3:

            
        indicador = 0
        for i in l1:
            for j in l2:
                nn = similar(i, j)
                if nn>=0.8:
                    indicador+=nn
        if indicador >= len(l1)-1:
            return True
        else:
            return False
    else:
        return False

In [132]:
trues = []   #nombre con coincidencias
falses = []  # nombre sin coincidencias

In [133]:
G = open("prob75.txt", "r")
linea = G.readline()
while len(linea)!=0:
    part = linea.strip().split(" | ")
    x, y, a, b = part
    TF = comparar2(x, y)
    if TF==True or float(b)>0.91:
        trues.append([x, y, a, b])
    else:
        falses.append([x, y, a, b])
    linea = G.readline()
G.close()

In [134]:
trues

[['a mushfiq mobarak', 'ahmed mushfiq mobarak', 'pmo232', '0.89474'],
 ['a. kerem co?ar', 'a. kerem cosar', 'pco958', '0.92857'],
 ['a. mushfiq mobarak', 'ahmed mushfiq mobarak', 'pmo232', '0.87179'],
 ['aaron s. edlin', 'aaron edlin', 'ped12', '0.88000'],
 ['aaron swoboda', 'aaron m. swoboda', 'psw23', '0.89655'],
 ['abel winn', 'abel m. winn', 'pwi202', '0.85714'],
 ['abhijit v. banerjee', 'abhijit banerjee', 'pba1710', '0.91429'],
 ['achyuta adhvaryu', 'achyuta r. adhvaryu', 'pad77', '0.91429'],
 ['adam ashcraft', 'adam b. ashcraft', 'pas35', '0.89655'],
 ['adam b. jaffe', 'adam jaffe', 'pja49', '0.86957'],
 ['adam copeland', 'adam matthew copeland', 'pco241', '0.76471'],
 ['adam m. guren', 'adam guren', 'pgu538', '0.86957'],
 ['adam m. lavecchia', 'adam m lavecchia', 'pla823', '0.96970'],
 ['adam m. rosen', 'adam michael rosen', 'pro441', '0.77419'],
 ['adnan q. khan', 'adnan khan', 'pkh414', '0.86957'],
 ['adnan q. khan', 'adnan khan', 'pkh420', '0.86957'],
 ['adriaan r. soetevent

In [135]:
T = []
for i in trues:
    T.append([i[0], i[1].title(), i[2]])
df_true = pd.DataFrame(T)
df_true.columns = ["author", "Name", "Short-Id"]
df_true

author                       Name Short-Id
0          a mushfiq mobarak      Ahmed Mushfiq Mobarak   pmo232
1             a. kerem co?ar             A. Kerem Cosar   pco958
2         a. mushfiq mobarak      Ahmed Mushfiq Mobarak   pmo232
3             aaron s. edlin                Aaron Edlin    ped12
4              aaron swoboda           Aaron M. Swoboda    psw23
...                      ...                        ...      ...
1329          yishay d. maoz          Yishay David Maoz   pma463
1330           young han lee             Young Hoon Lee   ple304
1331           young han lee             Yung-Chuan Lee   ple984
1332         yukiko u. saito         Yukiko Umeno Saito  psa1086
1333  zarek c. brot-goldberg  Zarek Chase Brot-Goldberg   pbr639

[1334 rows x 3 columns]

### Un total de 1334 autores con coincidencias
#### Aun asi, puede haber  un autor repetido por un determinado rango de error

### Se procede a merge-ar los DataFrames

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
df_eco= pd.read_csv('emetrp.csv', delimiter=';',encoding = "ISO-8859-1", error_bad_lines=False)
df_aer= pd.read_csv('aecrev.csv', delimiter=';',encoding = "ISO-8859-1", error_bad_lines=False)
df_jpe= pd.read_csv('jpolec.csv', delimiter=';',encoding = "ISO-8859-1", error_bad_lines=False)
df_qje= pd.read_csv('qjecon.csv', delimiter=';',encoding = "ISO-8859-1", error_bad_lines=False)
df_res= pd.read_csv('restud.csv', delimiter=';',encoding = "ISO-8859-1", error_bad_lines=False)

In [3]:
df=df_aer.append(df_eco,ignore_index=True)
df=df.append(df_qje,ignore_index=True)
df=df.append(df_jpe,ignore_index=True)
df=df.append(df_res,ignore_index=True)

In [4]:
df.drop(columns=['handle','title','vol','issue','page','end page'], inplace=True)

In [6]:
df

journal    year                   author
0        American Economic Review  2000.0          D. Gale Johnson
1        American Economic Review  2000.0        Dmitriy Stolyarov
2        American Economic Review  2000.0          Boyan Jovanovic
3        American Economic Review  2000.0        Eric S. Rosengren
4        American Economic Review  2000.0                 Joe Peek
...                           ...     ...                      ...
17026  Review of Economic Studies  2019.0      Jeong Ho (John) Kim
17027  Review of Economic Studies  2019.0          Eduardo Morales
17028  Review of Economic Studies  2019.0              Gloria Sheu
17029  Review of Economic Studies  2019.0            Andrés Zahler
17030  Review of Economic Studies  2019.0  Eduardo Souza-Rodrigues

[17031 rows x 3 columns]

In [7]:
df_repec=pd.read_csv('autores_repec.csv', delimiter=',')
df_repec

Name Short-Id
0              Arild Aakvik     paa1
1                 Eline Aas    paa10
2           Arnstein Aassve    paa11
3      Susan Ariel Aaronson    paa12
4           Daniel Aaronson    paa13
...                     ...      ...
70758   Eduardo Zylberstajn     pzy5
70759     Adam Zylbersztejn     pzy6
70760          Robert Zymek     pzy7
70761      Yanos Zylberberg     pzy8
70762       Benjamin Zycher     pzy9

[70763 rows x 2 columns]

In [8]:
df_true = pd.merge(df, df_repec, how = 'left',left_on ='author', right_on = 'Name')
df_true

journal    year                   author  \
0        American Economic Review  2000.0          D. Gale Johnson   
1        American Economic Review  2000.0        Dmitriy Stolyarov   
2        American Economic Review  2000.0          Boyan Jovanovic   
3        American Economic Review  2000.0        Eric S. Rosengren   
4        American Economic Review  2000.0                 Joe Peek   
...                           ...     ...                      ...   
17139  Review of Economic Studies  2019.0      Jeong Ho (John) Kim   
17140  Review of Economic Studies  2019.0          Eduardo Morales   
17141  Review of Economic Studies  2019.0              Gloria Sheu   
17142  Review of Economic Studies  2019.0            Andrés Zahler   
17143  Review of Economic Studies  2019.0  Eduardo Souza-Rodrigues   

                  Name Short-Id  
0      D. Gale Johnson   pjo254  
1                  NaN      NaN  
2      Boyan Jovanovic    pjo20  
3                  NaN      NaN  
4             Joe Peek    ppe90  
...                ...      ...  
17139              NaN      NaN  
17140              NaN      NaN  
17141      Gloria Sheu   psh447  
17142              NaN      NaN  
17143              NaN      NaN  

[17144 rows x 5 columns]

### Primer DataFrame creado con las coincidencias
##### No se requirio ningun filtro

In [143]:
df_final_1 = df_final.dropna()
df_final_1

journal    year             author  \
0        American Economic Review  2000.0    D. Gale Johnson   
2        American Economic Review  2000.0    Boyan Jovanovic   
4        American Economic Review  2000.0           Joe Peek   
5        American Economic Review  2000.0    Larry Samuelson   
6        American Economic Review  2000.0  George J. Mailath   
...                           ...     ...                ...   
17133  Review of Economic Studies  2019.0     Elisa Faraglia   
17134  Review of Economic Studies  2019.0      Albert Marcet   
17135  Review of Economic Studies  2019.0    Rigas Oikonomou   
17136  Review of Economic Studies  2019.0       Andrew Scott   
17141  Review of Economic Studies  2019.0        Gloria Sheu   

                    Name Short-Id  
0        D. Gale Johnson   pjo254  
2        Boyan Jovanovic    pjo20  
4               Joe Peek    ppe90  
5        Larry Samuelson    psa80  
6      George J. Mailath     pma6  
...                  ...      ...  
17133     Elisa Faraglia   pfa164  
17134      Albert Marcet   pma173  
17135    Rigas Oikonomou     poi7  
17136       Andrew Scott    psc65  
17141        Gloria Sheu   psh447  

[9522 rows x 5 columns]

### Segundo Dataframe con valores de repec no encontrados
#### Se procede a procesar usando un filtro de comparacion

In [144]:
df_final_0 = df_final[df_final.isna().any(axis=1)]
df_final_0 = df_final_0[['journal', 'year', 'author']]
df_final_0.sort_values(by='author')

journal    year                author
5934     American Economic Review  2013.0       ?kos Valentinyi
17022  Review of Economic Studies  2019.0     A Mushfiq Mobarak
4185     American Economic Review  2010.0  A. Joshua Strickland
7181     American Economic Review  2016.0        A. Kerem Co?ar
6771     American Economic Review  2015.0    A. Mushfiq Mobarak
...                           ...     ...                   ...
15911                         NaN     NaN                   NaN
15912                         NaN     NaN                   NaN
16692                         NaN     NaN                   NaN
16693                         NaN     NaN                   NaN
16694                         NaN     NaN                   NaN

[7622 rows x 3 columns]

In [145]:
autors_ = list(df_final_0["author"])
lista = []
for i in autors_:
    if type(i)==float:
        lista.append(i)
    else:
        lista.append(i.lower())
df_final_0['author'] = lista

In [146]:
df_final_0

journal    year                   author
1        American Economic Review  2000.0        dmitriy stolyarov
3        American Economic Review  2000.0        eric s. rosengren
7        American Economic Review  2000.0             avner shaked
8        American Economic Review  2000.0        david e. wildasin
10       American Economic Review  2000.0        thomas j. nechyba
...                           ...     ...                      ...
17138  Review of Economic Studies  2019.0            junehyuk jung
17139  Review of Economic Studies  2019.0      jeong ho (john) kim
17140  Review of Economic Studies  2019.0          eduardo morales
17142  Review of Economic Studies  2019.0            andrés zahler
17143  Review of Economic Studies  2019.0  eduardo souza-rodrigues

[7622 rows x 3 columns]

### Finalmente se obtiene el Dataframe filtrado
#### Aquellos autores con `NaN` en la columna `Name` no obtuvieron coincidencias

In [147]:
df_final_0_1 = pd.merge(df_final_0, df_true, how = 'left', left_on ='author', right_on = 'author')
df_final_0_1

journal    year                   author  \
0       American Economic Review  2000.0        dmitriy stolyarov   
1       American Economic Review  2000.0        eric s. rosengren   
2       American Economic Review  2000.0             avner shaked   
3       American Economic Review  2000.0        david e. wildasin   
4       American Economic Review  2000.0        thomas j. nechyba   
...                          ...     ...                      ...   
7812  Review of Economic Studies  2019.0            junehyuk jung   
7813  Review of Economic Studies  2019.0      jeong ho (john) kim   
7814  Review of Economic Studies  2019.0          eduardo morales   
7815  Review of Economic Studies  2019.0            andrés zahler   
7816  Review of Economic Studies  2019.0  eduardo souza-rodrigues   

                      Name Short-Id  
0     Dmitriy L. Stolyarov   pst142  
1           Eric Rosengren    pro88  
2                      NaN      NaN  
3           David Wildasin     pwi3  
4           Thomas Nechyba    pne28  
...                    ...      ...  
7812                   NaN      NaN  
7813                   NaN      NaN  
7814                   NaN      NaN  
7815                   NaN      NaN  
7816                   NaN      NaN  

[7817 rows x 5 columns]

In [149]:
df_final_final = df_final_1.append(df_final_0_1)
df_final_final

journal    year                   author  \
0       American Economic Review  2000.0          D. Gale Johnson   
2       American Economic Review  2000.0          Boyan Jovanovic   
4       American Economic Review  2000.0                 Joe Peek   
5       American Economic Review  2000.0          Larry Samuelson   
6       American Economic Review  2000.0        George J. Mailath   
...                          ...     ...                      ...   
7812  Review of Economic Studies  2019.0            junehyuk jung   
7813  Review of Economic Studies  2019.0      jeong ho (john) kim   
7814  Review of Economic Studies  2019.0          eduardo morales   
7815  Review of Economic Studies  2019.0            andrés zahler   
7816  Review of Economic Studies  2019.0  eduardo souza-rodrigues   

                   Name Short-Id  
0       D. Gale Johnson   pjo254  
2       Boyan Jovanovic    pjo20  
4              Joe Peek    ppe90  
5       Larry Samuelson    psa80  
6     George J. Mailath     pma6  
...                 ...      ...  
7812                NaN      NaN  
7813                NaN      NaN  
7814                NaN      NaN  
7815                NaN      NaN  
7816                NaN      NaN  

[17339 rows x 5 columns]

In [150]:
df_final_final.to_csv("autores_final.csv")